### DISCUSSION TITLE
Stripe: Grabbing the values from the form to pass to the checkout session

### DISCUSSION THREAD

**Discussion by Gabrielle-Tobermann** [nss-evening-cohort-14/evening-client-side#303] [Capstones!] [answered]
Gabby

other issues: no

Issue: Update: I am 99% of the way there with having a fully successful payment. Right now, I am being successfully redirected to stripe checkout page, and when that form is submitted, my payment is successfully processed (I can see it on my stripe dashboard). I am only using test data at the moment. I entered sample values in the line_items key in the object that is passed to the checkout session in my create-checkout file (which creates a checkout session when the order form is submitted. The last blocker I have with this is grabbing the values of the item from the form, instead of the sample values. I am not sure how to pass the values to a non react component. I would like to be able to have several items per order, but it looks like I will have to do one item per order. 

I looked back at the Netlify docs: 
https://docs.netlify.com/functions/build-with-javascript/
This is the tutorial I am following: 
https://www.netlify.com/blog/2020/04/13/learn-how-to-accept-money-on-jamstack-sites-in-38-minutes/
https://stackoverflow.com/questions/61634308/alternative-for-document-queryselector-in-react-js
My code:
checkout session file:
https://github.com/Gabrielle-Tobermann/Overture/blob/read-create-orders/functions/create-checkout.js
order form:
https://github.com/Gabrielle-Tobermann/Overture/blob/read-create-orders/src/components/OrderForm.js

**Comment by drteresavasquez** [comment, marked as answer]
@Gabrielle-Tobermann did you look at the other resources I shared outside of Netlify functions? You should be able to try and incorporate some of those into your send. The video is helpful as well.

**Reply by Gabrielle-Tobermann** [reply]
I will look through those again. I also found something on the netlify forum that mentions that you need an html form that matches exactly your form https://answers.netlify.com/t/support-guide-my-form-renders-with-javascript-but-wont-work-properly/87/7

**Reply by Gabrielle-Tobermann** [reply]
So I added an html form to match my react form, just like the docs said. I am getting closer to grabbing the values, but I am still getting an 500 error, still at the same place (in the line_items key, in the object I am passing to  stripe.checkout.sessions.create. This for the name of the item, I am passing product.itemID, but is is returning cannot read property itemID of undefined. I made sure the name attribute in both forms match. 

My create-checkout file: 
https://github.com/Gabrielle-Tobermann/Overture/blob/read-create-orders/functions/create-checkout.js
form: 
https://github.com/Gabrielle-Tobermann/Overture/blob/read-create-orders/src/components/OrderForm.js
https://github.com/Gabrielle-Tobermann/Overture/blob/read-create-orders/src/orderForm.html

Helpful link from Netlify for the form stuff: 
https://www.netlify.com/blog/2017/07/20/how-to-integrate-netlifys-form-handling-in-a-react-app/#form-handling-with-a-stateless-react-form
First tutorial about netlify severless functions: 
https://www.netlify.com/blog/2020/04/13/learn-how-to-accept-money-on-jamstack-sites-in-38-minutes/

**Reply by drteresavasquez** [reply]
You need the form located here with the encode function: Form Handling with a Stateful React Form. You have to scroll a good way down the page to see it. Once you have that set up in the form you have, we can make a fetch but part of me feels like we don't have to do all of that. 

We can stringify the state object to send to your order processing file.

**Reply by drteresavasquez** [reply]
For specifics, this is the form you will need to use. Once you finish up MVP, we can revisit. Closing this until then.

```
const encode = (data) => {
    return Object.keys(data)
        .map(key => encodeURIComponent(key) + "=" + encodeURIComponent(data[key]))
        .join("&");
  }

  class ContactForm extends React.Component {
    constructor(props) {
      super(props);
      this.state = { name: "", email: "", message: "" };
    }

    /* Here’s the juicy bit for posting the form submission */

    handleSubmit = e => {
      fetch("/", {
        method: "POST",
        headers: { "Content-Type": "application/x-www-form-urlencoded" },
        body: encode({ "form-name": "contact", ...this.state })
      })
        .then(() => alert("Success!"))
        .catch(error => alert(error));

      e.preventDefault();
    };

    handleChange = e => this.setState({ [e.target.name]: e.target.value });

    render() {
      const { name, email, message } = this.state;
      return (
        <form onSubmit={this.handleSubmit}>
          <p>
            <label>
              Your Name: <input type="text" name="name" value={name} onChange={this.handleChange} />
            </label>
          </p>
          <p>
            <label>
              Your Email: <input type="email" name="email" value={email} onChange={this.handleChange} />
            </label>
          </p>
          <p>
            <label>
              Message: <textarea name="message" value={message} onChange={this.handleChange} />
            </label>
          </p>
          <p>
            <button type="submit">Send</button>
          </p>
        </form>
      );
    }
  }
```
